DOC: update the pandas.Series/DataFrame.interpolate docstring

math-and-data opened the pull request:
- [x] PR title is "DOC: update the <your-function-or-method> docstring"
- [ ] The validation script passes: `scripts/validate_docstrings.py <your-function-or-method>`
- [x] The PEP8 style check passes: `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] The html version looks good: `python doc/make.py --single <your-function-or-method>`
- [ ] It has been proofread on language by another sprint participant

Errors in the validation script:
- Formatting of the 'method' param is not right (not sure how to break option list properly into multiple lines)
- kwargs (to ignore)
- Formatting compaints are due to extra line for "New in version ..."
```################################################################################
################### Docstring (pandas.DataFrame.interpolate) ###################
################################################################################

Interpolate values according to different methods.

Please note that only ``method='linear'`` is supported for
DataFrames/Series with a MultiIndex.

Parameters
----------
method : {'linear', 'time', 'index', 'values', 'nearest', 'zero',
          'slinear', 'quadratic', 'cubic', 'barycentric', 'krogh',
          'polynomial', 'spline', 'piecewise_polynomial', 'pad',
          'from_derivatives', 'pchip', 'akima'}, default 'linear'
    Interpolation technique to use.

    * 'linear': Ignore the index and treat the values as equally
      spaced. This is the only method supported on MultiIndexes.
      Default.
    * 'time': Interpolation works on daily and higher resolution
      data to interpolate given length of interval.
    * 'index', 'values': use the actual numerical values of the index.
    * 'pad': Fill in NaNs using existing values.
    * 'nearest', 'zero', 'slinear', 'quadratic', 'cubic',
      'barycentric', 'polynomial': Passed to
      ``scipy.interpolate.interp1d``. Both 'polynomial' and 'spline'
      require that you also specify an `order` (int),
      e.g. df.interpolate(method='polynomial', order=4).
      These use the actual numerical values of the index.
    * 'krogh', 'piecewise_polynomial', 'spline', 'pchip', 'akima':
      Wrappers around the scipy interpolation methods of
      similar names. These use the actual numerical values of the
      index. For more information on their behavior, see the
      `scipy documentation
      <http://docs.scipy.org/doc/scipy/reference/interpolate.html#univariate-interpolation>`__
      and `tutorial documentation
      <http://docs.scipy.org/doc/scipy/reference/tutorial/interpolate.html>`__.
    * 'from_derivatives': Refers to
      ``scipy.intrepolate.BPoly.from_derivatives`` which
      replaces 'piecewise_polynomial' interpolation method in
      scipy 0.18.

    .. versionadded:: 0.18.1

       Added support for the 'akima' method
       Added interpolate method 'from_derivatives' which replaces
       'piecewise_polynomial' in scipy 0.18; backwards-compatible with
       scipy < 0.18

axis : {0, 1}, default 0
    Axis to interpolate along.

    * 0: Fill column-by-column.
    * 1: Fill row-by-row.
limit : int, default None
    Maximum number of consecutive NaNs to fill. Must be greater than 0.
inplace : bool, default False
    Update the data in place if possible.
limit_direction : {'forward', 'backward', 'both'}, default 'forward'
    If limit is specified, consecutive NaNs will be filled in this
    direction.
limit_area : {'inside', 'outside'}, default None
    If limit is specified, consecutive NaNs will be filled with this
    restriction.

    * None: No fill restriction (default).
    * 'inside': Only fill NaNs surrounded by valid values
      (interpolate).
    * 'outside': Only fill NaNs outside valid values (extrapolate).

    .. versionadded:: 0.21.0

downcast : optional, 'infer' or None, defaults to None
    Downcast dtypes if possible.
kwargs
    Keyword arguments to pass on to the interpolating function.

Returns
-------
Series or DataFrame
    Same-shape object interpolated at the NaN values

See Also
--------
replace : replace a value
fillna : fill missing values

Examples
--------

Filling in NaNs in a Series via linear interpolation.

>>> ser = pd.Series([0, 1, np.nan, 3])
>>> ser.interpolate()
0    0.0
1    1.0
2    2.0
3    3.0
dtype: float64

Filling in NaNs in a Series by padding, but filling at most two
consecutive NaN at a time.

>>> ser = pd.Series([np.nan, "single_one", np.nan,
...                  "fill_two_more", np.nan, np.nan, np.nan,
...                  4.71, np.nan])
>>> ser
0              NaN
1       single_one
2              NaN
3    fill_two_more
4              NaN
5              NaN
6              NaN
7             4.71
8              NaN
dtype: object
>>> ser.interpolate(method='pad', limit=2)
0              NaN
1       single_one
2       single_one
3    fill_two_more
4    fill_two_more
5    fill_two_more
6              NaN
7             4.71
8             4.71
dtype: object

Create a DataFrame with missing values.

>>> df = pd.DataFrame([[0,1,2,0,4],[1,2,3,-1,8],
...                    [2,3,4,-2,12],[3,4,5,-3,16]],
...                   columns=['a', 'b', 'c', 'd', 'e'])
>>> df
   a  b  c  d   e
0  0  1  2  0   4
1  1  2  3 -1   8
2  2  3  4 -2  12
3  3  4  5 -3  16
>>> df.loc[3,'a'] = np.nan
>>> df.loc[0,'b'] = np.nan
>>> df.loc[1,'d'] = np.nan
>>> df.loc[2,'d'] = np.nan
>>> df.loc[1,'e'] = np.nan
>>> df
     a    b  c    d     e
0  0.0  NaN  2  0.0   4.0
1  1.0  2.0  3  NaN   NaN
2  2.0  3.0  4  NaN  12.0
3  NaN  4.0  5 -3.0  16.0

Fill the DataFrame forward (that is, going down) along each column.
Note how the last entry in column `a` is interpolated differently
(because there is no entry after it to use for interpolation).
Note how the first entry in column `b` remains NA (because there
is no entry befofe it to use for interpolation).

>>> df.interpolate(method='linear', limit_direction='forward', axis=0)
     a    b  c    d     e
0  0.0  NaN  2  0.0   4.0
1  1.0  2.0  3 -1.0   8.0
2  2.0  3.0  4 -2.0  12.0
3  2.0  4.0  5 -3.0  16.0

################################################################################
################################## Validation ##################################
################################################################################

Errors found:
        Errors in parameters section
                Parameter "method" description should start with capital letter
                Parameter "method" description should finish with "."
                Parameter "limit_area" description should finish with "."
                Parameter "kwargs" has no type
```

math-and-data committed:
DOC: update pd.Series/DataFrame.interpolate

math-and-data committed:
DOC: update pd.Series/DataFrame.interpolate, removed whitespace

math-and-data committed:
DOC: update pd.Series/DataFrame.interpolate, new allowed value for input param

WillAyd commented on pandas/core/generic.py:
```diff
@@ -5292,46 +5295,120 @@ def replace(self, to_replace=None, value=None, inplace=False, limit=None,
+
+            * 0: Fill column-by-column.
+            * 1: Fill row-by-row.
+        limit : int, default None
```
Since `None` isn't an explicitly used value better to say optional

WillAyd commented on pandas/core/generic.py:
```diff
@@ -5292,46 +5295,120 @@ def replace(self, to_replace=None, value=None, inplace=False, limit=None,
+        inplace : bool, default False
+            Update the data in place if possible.
         limit_direction : {'forward', 'backward', 'both'}, default 'forward'
+            If limit is specified, consecutive NaNs will be filled in this
```
Put back ticks around \`NaN\`

WillAyd commented on pandas/core/generic.py:
```diff
@@ -5292,46 +5295,120 @@ def replace(self, to_replace=None, value=None, inplace=False, limit=None,
         limit_direction : {'forward', 'backward', 'both'}, default 'forward'
+            If limit is specified, consecutive NaNs will be filled in this
+            direction.
         limit_area : {'inside', 'outside'}, default None
```
{\`None\`, 'inside', 'outside'}

WillAyd commented on pandas/core/generic.py:
```diff
@@ -5292,46 +5295,120 @@ def replace(self, to_replace=None, value=None, inplace=False, limit=None,
         downcast : optional, 'infer' or None, defaults to None
             Downcast dtypes if possible.
-        kwargs : keyword arguments to pass on to the interpolating function.
+        kwargs
```
**kwargs

WillAyd commented on pandas/core/generic.py:
```diff
@@ -5292,46 +5295,120 @@ def replace(self, to_replace=None, value=None, inplace=False, limit=None,
         See Also
         --------
-        reindex, replace, fillna
+        replace : replace a value
```
See comments above - so much can be added here

math-and-data replied:
done

WillAyd commented on pandas/core/generic.py:
```diff
@@ -5292,46 +5295,120 @@ def replace(self, to_replace=None, value=None, inplace=False, limit=None,
         --------
 
-        Filling in NaNs
+        Filling in NaNs in a Series via linear interpolation.
```
:class:\`~pandas.Series\`

WillAyd commented on pandas/core/generic.py:
```diff
@@ -5292,46 +5295,120 @@ def replace(self, to_replace=None, value=None, inplace=False, limit=None,
-        1    1
-        2    2
-        3    3
+        >>> ser = pd.Series([0, 1, np.nan, 3])
```
Convention here is `s =` instead of `ser =`

WillAyd commented on pandas/core/generic.py:
```diff
@@ -5292,46 +5295,120 @@ def replace(self, to_replace=None, value=None, inplace=False, limit=None,
+        >>> ser = pd.Series([np.nan, "single_one", np.nan,
+        ...                  "fill_two_more", np.nan, np.nan, np.nan,
+        ...                  4.71, np.nan])
+        >>> ser
```
To save space you don't need to print the `Series` here - should be straightforward based off the constructor directly above it

math-and-data replied:
I did not include the print for the very small series example (it was straightforward to see), but I'd like to keep this longer one if that's alright - it was encouraged so the differences can be spotted easier.

WillAyd commented on pandas/core/generic.py:
```diff
@@ -5292,46 +5295,120 @@ def replace(self, to_replace=None, value=None, inplace=False, limit=None,
+
+        Create a DataFrame with missing values.
+
+        >>> df = pd.DataFrame([[0,1,2,0,4],[1,2,3,-1,8],
```
Why not just construct with the missing values?

math-and-data replied:
Mainly so people can see the "expected" Interpolation (I tried to have a pattern column-wise) and they can compare it with what actually happens, e.g. with lin. Interpolation (especially if the last entry is an NA)

WillAyd commented on pandas/core/generic.py:
```diff
@@ -5292,46 +5295,120 @@ def replace(self, to_replace=None, value=None, inplace=False, limit=None,
+        Fill the DataFrame forward (that is, going down) along each column.
+        Note how the last entry in column `a` is interpolated differently
+        (because there is no entry after it to use for interpolation).
+        Note how the first entry in column `b` remains NA (because there
```
`NaN`

WillAyd commented:
Lot's of comments but just wanted to say nice job! This is one of the tougher docstrings

math-and-data commented:
Thank you for the thorough review, @WillAyd 

I believe the errors below can be ignored, because they relate to known issues (**kwargs, .. versionadded::, etc.)
```
################################################################################
################### Docstring (pandas.DataFrame.interpolate) ###################
################################################################################

Interpolate values according to different methods.

Please note that only ``method='linear'`` is supported for
DataFrames/Series with a MultiIndex.

Parameters
----------
method : {'linear', 'time', 'index', 'values', 'nearest', 'zero',
          'slinear', 'quadratic', 'cubic', 'barycentric', 'krogh',
          'polynomial', 'spline', 'piecewise_polynomial', 'pad',
          'from_derivatives', 'pchip', 'akima'}
    Interpolation technique to use.

    * 'linear': Ignore the index and treat the values as equally
      spaced. This is the only method supported on MultiIndexes.
    * 'time': Works on daily and higher resolution
      data to interpolate given length of interval.
    * 'index', 'values': use the actual numerical values of the index.
    * 'pad': Fill in `NaN`s using existing values.
    * 'nearest', 'zero', 'slinear', 'quadratic', 'cubic',
      'barycentric', 'polynomial': Passed to
      ``scipy.interpolate.interp1d``. Both 'polynomial' and 'spline'
      require that you also specify an `order` (int),
      e.g. df.interpolate(method='polynomial', order=4).
      These use the actual numerical values of the index.
    * 'krogh', 'piecewise_polynomial', 'spline', 'pchip', 'akima':
      Wrappers around the scipy interpolation methods of
      similar names. See `Notes`.
    * 'from_derivatives': Refers to
      ``scipy.interpolate.BPoly.from_derivatives`` which
      replaces 'piecewise_polynomial' interpolation method in
      scipy 0.18.

    .. versionadded:: 0.18.1

       Added support for the 'akima' method
       Added interpolate method 'from_derivatives' which replaces
       'piecewise_polynomial' in scipy 0.18; backwards-compatible with
       scipy < 0.18

axis : {0 or 'index', 1 or 'columns', None}, default None
    Axis to interpolate along.
limit : int, optional
    Maximum number of consecutive `NaN`s to fill. Must be greater than 0.
inplace : bool, default False
    Update the data in place if possible.
limit_direction : {'forward', 'backward', 'both'}, default 'forward'
    If limit is specified, consecutive `NaN`s will be filled in this
    direction.
limit_area : {`None`, 'inside', 'outside'}
    If limit is specified, consecutive `NaN`s will be filled with this
    restriction.

    * None: No fill restriction (default).
    * 'inside': Only fill `NaN`s surrounded by valid values
      (interpolate).
    * 'outside': Only fill `NaN`s outside valid values (extrapolate).

    .. versionadded:: 0.21.0

downcast : optional, 'infer' or None, defaults to None
    Downcast dtypes if possible.
**kwargs
    Keyword arguments to pass on to the interpolating function.

Returns
-------
Series or DataFrame
    Same-shape object interpolated at the NaN values

See Also
--------
replace : replace a value
fillna : fill missing values
scipy.interpolate.Akima1DInterpolator : piecewise cubic polynomials
    (Akima interpolator)
scipy.interpolate.BPoly.from_derivatives : piecewise polynomial in the
    Bernstein basis
scipy.interpolate.interp1d : interpolate a 1-D function
scipy.interpolate.KroghInterpolator : interpolate polynomial (Krogh
    interpolator)
scipy.interpolate.PchipInterpolator : PCHIP 1-d monotonic cubic
    interpolation
scipy.interpolate.CubicSpline : cubic spline data interpolator

Notes
-----
If the selected `method` is one of 'krogh', 'piecewise_polynomial',
'spline', 'pchip', 'akima':
They are wrappers around the scipy interpolation methods of similar
names. These use the actual numerical values of the index.
For more information on their behavior, see the
`scipy documentation
<http://docs.scipy.org/doc/scipy/reference/interpolate.html#univariate-interpolation>`__
and `tutorial documentation
<http://docs.scipy.org/doc/scipy/reference/tutorial/interpolate.html>`__.

Examples
--------

Filling in `NaNs` in a :class:`~pandas.Series` via linear interpolation.

>>> s = pd.Series([0, 1, np.nan, 3])
>>> s.interpolate()
0    0.0
1    1.0
2    2.0
3    3.0
dtype: float64

Filling in `NaN`s in a Series by padding, but filling at most two
consecutive `NaN` at a time.

>>> s = pd.Series([np.nan, "single_one", np.nan,
...                  "fill_two_more", np.nan, np.nan, np.nan,
...                  4.71, np.nan])
>>> s
0              NaN
1       single_one
2              NaN
3    fill_two_more
4              NaN
5              NaN
6              NaN
7             4.71
8              NaN
dtype: object
>>> s.interpolate(method='pad', limit=2)
0              NaN
1       single_one
2       single_one
3    fill_two_more
4    fill_two_more
5    fill_two_more
6              NaN
7             4.71
8             4.71
dtype: object

Create a :class:`~pandas.DataFrame` with missing values to fill it with diffferent
methods.

>>> df = pd.DataFrame([[0,1,2,0,4],[1,2,3,-1,8],
...                    [2,3,4,-2,12],[3,4,5,-3,16]],
...                   columns=['a', 'b', 'c', 'd', 'e'])
>>> df
   a  b  c  d   e
0  0  1  2  0   4
1  1  2  3 -1   8
2  2  3  4 -2  12
3  3  4  5 -3  16
>>> df.loc[3,'a'] = np.nan
>>> df.loc[0,'b'] = np.nan
>>> df.loc[1,'d'] = np.nan
>>> df.loc[2,'d'] = np.nan
>>> df.loc[1,'e'] = np.nan
>>> df
     a    b  c    d     e
0  0.0  NaN  2  0.0   4.0
1  1.0  2.0  3  NaN   NaN
2  2.0  3.0  4  NaN  12.0
3  NaN  4.0  5 -3.0  16.0

Fill the DataFrame forward (that is, going down) along each column.
Note how the last entry in column `a` is interpolated differently
(because there is no entry after it to use for interpolation).
Note how the first entry in column `b` remains `NaN` (because there
is no entry befofe it to use for interpolation).

>>> df.interpolate(method='linear', limit_direction='forward', axis=0)
     a    b  c    d     e
0  0.0  NaN  2  0.0   4.0
1  1.0  2.0  3 -1.0   8.0
2  2.0  3.0  4 -2.0  12.0
3  2.0  4.0  5 -3.0  16.0

################################################################################
################################## Validation ##################################
################################################################################

Errors found:
        Errors in parameters section
                Parameters {'kwargs'} not documented
                Unknown parameters {'**kwargs'}
                Parameter "method" description should start with capital letter
                Parameter "method" description should finish with "."
                Parameter "limit_area" description should finish with "."
                Parameter "**kwargs" has no type
```

math-and-data committed:
DOC: pandas.DataFrame.interpolate incorporated recommended channges

math-and-data committed:
DOC: pandas.DataFrame.interpolnew examples

math-and-data committed:
DOC: pandas.DataFrame.interpol example update

WillAyd commented on pandas/core/generic.py:
```diff
@@ -5257,32 +5257,29 @@ def replace(self, to_replace=None, value=None, inplace=False, limit=None,
+            * 'pad': Fill in NaNs using existing values.
+            * 'nearest', 'zero', 'slinear', 'quadratic', 'cubic', 'spline',
+              'barycentric', 'polynomial': Passed to
               ``scipy.interpolate.interp1d``. Both 'polynomial' and 'spline'
```
I would do the same thing here you did for 'krogh' and move some of the implementation details down to the Notes section

WillAyd commented on pandas/core/generic.py:
```diff
@@ -5257,32 +5257,29 @@ def replace(self, to_replace=None, value=None, inplace=False, limit=None,
-              <http://docs.scipy.org/doc/scipy/reference/tutorial/interpolate.html>`__
-            * 'from_derivatives' refers to BPoly.from_derivatives which
+            * 'krogh', 'piecewise_polynomial', 'spline', 'pchip', 'akima':
+              Wrappers around the scipy interpolation methods of similar
```
Use SciPy instead of scipy when referring to the package outside of code (couple other places this pops up)

math-and-data replied:
sure

WillAyd commented on pandas/core/generic.py:
```diff
@@ -5291,47 +5288,162 @@ def replace(self, to_replace=None, value=None, inplace=False, limit=None,
             direction.
-        inplace : bool, default False
-            Update the NDFrame in place if possible.
+        limit_area : {`None`, 'inside', 'outside'}
```
Add ", default None" to the end here and remove the comment about it being the default below

WillAyd commented on pandas/core/generic.py:
```diff
@@ -5291,47 +5288,162 @@ def replace(self, to_replace=None, value=None, inplace=False, limit=None,
+            * None: No fill restriction (default).
+            * 'inside': Only fill NaNs surrounded by valid values
+              (interpolate).
+            * 'outside': Only fill NaNs outside valid values (extrapolate).
```
Would be good to add an example for 'outside'

WillAyd commented on pandas/core/generic.py:
```diff
@@ -5291,47 +5288,162 @@ def replace(self, to_replace=None, value=None, inplace=False, limit=None,
+
+        Notes
+        -----
+        If the selected `method` is one of 'krogh', 'piecewise_polynomial',
```
"The 'krogh', 'piecewise_polynomial', ... methods are wrappers around the respective SciPy implementations" would be better wording

WillAyd commented on pandas/core/generic.py:
```diff
@@ -5291,47 +5288,162 @@ def replace(self, to_replace=None, value=None, inplace=False, limit=None,
+        If the selected `method` is one of 'krogh', 'piecewise_polynomial',
+        'spline', 'pchip', 'akima':
+        They are wrappers around the scipy interpolation methods of similar
+        names. These use the actual numerical values of the index.
```
What does "These use the actual numerical values of the index" mean?

math-and-data replied:
These use the ~~actual~~ numerical values of the index. Better grammar?

WillAyd commented on pandas/core/generic.py:
```diff
@@ -5291,47 +5288,162 @@ def replace(self, to_replace=None, value=None, inplace=False, limit=None,
+        3    8.000000
+        dtype: float64
+
+        Create a :class:`~pandas.DataFrame` with missing values to fill it
```
You are explaining here what the below code is going to do, but not really saying what it's important. Would be better worded as "Interpolation can also be applied to DataFrames" or something to the effect

WillAyd commented on pandas/core/generic.py:
```diff
@@ -5291,47 +5288,162 @@ def replace(self, to_replace=None, value=None, inplace=False, limit=None,
+        Create a :class:`~pandas.DataFrame` with missing values to fill it
+        with diffferent methods.
 
+        >>> df = pd.DataFrame([[0,1,2,0,4],[1,2,3,-1,8],
```
Thought I had this comment before but just use the NA values in your constructor - no reason to instantiate the DataFrame with values and then assign them missing values after the fact.

Also make sure you put a space after every comma

math-and-data replied:
I changed it so one can see how the columns get created - and we have linear values in 3 columns and quadratic on the 4th.

WillAyd commented on pandas/core/generic.py:
```diff
@@ -5291,47 +5288,162 @@ def replace(self, to_replace=None, value=None, inplace=False, limit=None,
+
+        Fill the DataFrame forward (that is, going down) along each column.
+        Note how the last entry in column `a` is interpolated differently
+        (because there is no entry after it to use for interpolation).
```
Don't need the parentheses here (nor on the next line)

WillAyd commented on pandas/core/generic.py:
```diff
@@ -5291,47 +5288,162 @@ def replace(self, to_replace=None, value=None, inplace=False, limit=None,
         -------
-        Series or DataFrame of same shape interpolated at the NaNs
+        Series or DataFrame
+            Same-shape object interpolated at the NaN values
```
For the description here say "Returns the same object type as the caller" - that wording has been used by a few other PRs so just want to be consistent

math-and-data committed:
DOC> pandas.DataFrame.interpolate - implemented feedback

pep8speaks commented:
Hello @math-and-data! Thanks for updating the PR.

 - In the file [`pandas/core/generic.py`](https://github.com/pandas-dev/pandas/blob/d64f5f38aa940e484c441f7fadaadfa14ba0838f/pandas/core/generic.py), following are the PEP8 issues :

> [Line 5408:1](https://github.com/pandas-dev/pandas/blob/d64f5f38aa940e484c441f7fadaadfa14ba0838f/pandas/core/generic.py#L5408): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

math-and-data committed:
DOC: pandas.DataFrame.interpolate - implemented feedback

math-and-data committed:
DOC: pandas.DataFrame.interpolate - sneaky whitespace

math-and-data commented:
- [ ] docstring validation issues below, should be known kwarg and .
```
################################################################################
################### Docstring (pandas.DataFrame.interpolate) ###################
################################################################################

Interpolate values according to different methods.

Please note that only ``method='linear'`` is supported for
DataFrames/Series with a MultiIndex.

Parameters
----------
method : {'linear', 'time', 'index', 'values', 'nearest', 'zero',
          'slinear', 'quadratic', 'cubic', 'barycentric', 'krogh',
          'polynomial', 'spline', 'piecewise_polynomial', 'pad',
          'from_derivatives', 'pchip', 'akima'}
    Interpolation technique to use.

    * 'linear': Ignore the index and treat the values as equally
      spaced. This is the only method supported on MultiIndexes.
    * 'time': Works on daily and higher resolution
      data to interpolate given length of interval.
    * 'index', 'values': use the actual numerical values of the index.
    * 'pad': Fill in NaNs using existing values.
    * 'nearest', 'zero', 'slinear', 'quadratic', 'cubic', 'spline',
      'barycentric', 'polynomial': Passed to
      ``scipy.interpolate.interp1d``. Both 'polynomial' and 'spline'
      require that you also specify an `order` (int),
      e.g. df.interpolate(method='polynomial', order=4).
      These use the numerical values of the index.
    * 'krogh', 'piecewise_polynomial', 'spline', 'pchip', 'akima':
      Wrappers around the SciPy interpolation methods of similar
      names. See `Notes`.
    * 'from_derivatives': Refers to
      ``scipy.interpolate.BPoly.from_derivatives`` which
      replaces 'piecewise_polynomial' interpolation method in
      scipy 0.18.

    .. versionadded:: 0.18.1

       Added support for the 'akima' method
       Added interpolate method 'from_derivatives' which replaces
       'piecewise_polynomial' in SciPy 0.18; backwards-compatible with
       SciPy < 0.18

axis : {0 or 'index', 1 or 'columns', None}, default None
    Axis to interpolate along.
limit : int, optional
    Maximum number of consecutive NaNs to fill. Must be greater than
    0.
inplace : bool, default False
    Update the data in place if possible.
limit_direction : {'forward', 'backward', 'both'}, default 'forward'
    If limit is specified, consecutive NaNs will be filled in this
    direction.
limit_area : {`None`, 'inside', 'outside'}
    If limit is specified, consecutive NaNs will be filled with this
    restriction.

    * None: No fill restriction.
    * 'inside': Only fill NaNs surrounded by valid values
      (interpolate).
    * 'outside': Only fill NaNs outside valid values (extrapolate).

    .. versionadded:: 0.21.0

downcast : optional, 'infer' or None, defaults to None
    Downcast dtypes if possible.
**kwargs
    Keyword arguments to pass on to the interpolating function.

Returns
-------
Series or DataFrame
    Returns the same object type as the caller, interpolated at
    some or all `NaN` values

See Also
--------
replace : replace a value
fillna : fill missing values
scipy.interpolate.Akima1DInterpolator : piecewise cubic polynomials
    (Akima interpolator)
scipy.interpolate.BPoly.from_derivatives : piecewise polynomial in the
    Bernstein basis
scipy.interpolate.interp1d : interpolate a 1-D function
scipy.interpolate.KroghInterpolator : interpolate polynomial (Krogh
    interpolator)
scipy.interpolate.PchipInterpolator : PCHIP 1-d monotonic cubic
    interpolation
scipy.interpolate.CubicSpline : cubic spline data interpolator

Notes
-----
The 'krogh', 'piecewise_polynomial', 'spline', 'pchip' and 'akima'
methods are wrappers around the respective SciPy implementations of
similar names. These use the actual numerical values of the index.
For more information on their behavior, see the
`SciPy documentation
<http://docs.scipy.org/doc/scipy/reference/interpolate.html#univariate-interpolation>`__
and `SciPy tutorial
<http://docs.scipy.org/doc/scipy/reference/tutorial/interpolate.html>`__.

Examples
--------

Filling in `NaN` in a :class:`~pandas.Series` via linear
interpolation.

>>> s = pd.Series([0, 1, np.nan, 3])
>>> s.interpolate()
0    0.0
1    1.0
2    2.0
3    3.0
dtype: float64

Filling in `NaN` in a Series by padding, but filling at most two
consecutive `NaN` at a time.

>>> s = pd.Series([np.nan, "single_one", np.nan,
...                "fill_two_more", np.nan, np.nan, np.nan,
...                4.71, np.nan])
>>> s
0              NaN
1       single_one
2              NaN
3    fill_two_more
4              NaN
5              NaN
6              NaN
7             4.71
8              NaN
dtype: object
>>> s.interpolate(method='pad', limit=2)
0              NaN
1       single_one
2       single_one
3    fill_two_more
4    fill_two_more
5    fill_two_more
6              NaN
7             4.71
8             4.71
dtype: object

Filling in `NaN` in a Series via polynomial interpolation or splines:
Both `polynomial` and `spline` methods require that you also specify
an `order` (int).

>>> s = pd.Series([0, 2, np.nan, 8])
>>> s.interpolate(method='polynomial', order=2)
0    0.000000
1    2.000000
2    4.666667
3    8.000000
dtype: float64

Filling in `NaN` in a :class:`~pandas.DataFrame` via linear
interpolation.

>>> df = pd.DataFrame({'a': range(0,4),
...                    'b': range(1,5),
...                    'c': range(-1, -5, -1),
...                    'd': [x**2 for x in range(1,5)]})
>>> df
   a  b  c   d
0  0  1 -1   1
1  1  2 -2   4
2  2  3 -3   9
3  3  4 -4  16
>>> df.loc[1,'a'] = np.nan
>>> df.loc[3,'a'] = np.nan
>>> df.loc[0,'b'] = np.nan
>>> df.loc[1,'c'] = np.nan
>>> df.loc[2,'c'] = np.nan
>>> df.loc[1,'d'] = np.nan
>>> df
     a    b    c     d
0  0.0  NaN -1.0   1.0
1  NaN  2.0  NaN   NaN
2  2.0  3.0  NaN   9.0
3  NaN  4.0 -4.0  16.0

Fill the DataFrame forward (that is, going down) along each column.
Note how the last entry in column `a` is interpolated differently,
because there is no entry after it to use for interpolation.
Note how the first entry in column `b` remains `NaN`, because there
is no entry befofe it to use for interpolation.

>>> df.interpolate(method='linear', limit_direction='forward', axis=0)
     a    b    c     d
0  0.0  NaN -1.0   1.0
1  1.0  2.0 -2.0   5.0
2  2.0  3.0 -3.0   9.0
3  2.0  4.0 -4.0  16.0

>>> df['d'].interpolate(method='polynomial', order=2)
0     1.0
1     4.0
2     9.0
3    16.0
Name: d, dtype: float64

################################################################################
################################## Validation ##################################
################################################################################

Errors found:
        Errors in parameters section
                Parameters {'kwargs'} not documented
                Unknown parameters {'**kwargs'}
                Parameter "method" description should start with capital letter
                Parameter "method" description should finish with "."
                Parameter "limit_area" description should finish with "."
                Parameter "**kwargs" has no type
```

datapythonista committed:
Minor fixes to interpolate docstring

datapythonista committed:
Merging from master

datapythonista committed:
Last minor fixes

datapythonista reviewed (approved):
@WillAyd if you don't mind reviewing this one too. Made few minor changes like pep8 of the doctest, the type of method couldn't be all options, as sphinx do not let parameter types be multiline and things like this.

Thanks for the docstring @math-and-data, really good work. And sorry for the long wait.

WillAyd commented on pandas/core/generic.py:
```diff
@@ -6081,89 +6081,191 @@ def replace(self, to_replace=None, value=None, inplace=False, limit=None,
+            If limit is specified, consecutive NaNs will be filled with this
+            restriction.
+
+            * None: No fill restriction.
```
Maybe double backticks here to render literal value?

WillAyd commented on pandas/core/generic.py:
```diff
@@ -6081,89 +6081,191 @@ def replace(self, to_replace=None, value=None, inplace=False, limit=None,
-        Series or DataFrame of same shape interpolated at the NaNs
+        Series or DataFrame
+            Returns the same object type as the caller, interpolated at
+            some or all `NaN` values
```
Double ticks

WillAyd commented on pandas/core/generic.py:
```diff
@@ -6081,89 +6081,191 @@ def replace(self, to_replace=None, value=None, inplace=False, limit=None,
         --------
-
-        Filling in NaNs
+        Filling in `NaN` in a :class:`~pandas.Series` via linear
```
Double

WillAyd commented on pandas/core/generic.py:
```diff
@@ -6081,89 +6081,191 @@ def replace(self, to_replace=None, value=None, inplace=False, limit=None,
+        3    3.0
         dtype: float64
 
+        Filling in `NaN` in a Series by padding, but filling at most two
```
Double ticks (here and next line)

WillAyd commented on pandas/core/generic.py:
```diff
@@ -6081,89 +6081,191 @@ def replace(self, to_replace=None, value=None, inplace=False, limit=None,
+        8             4.71
+        dtype: object
+
+        Filling in `NaN` in a Series via polynomial interpolation or splines:
```
Double backtick

WillAyd commented on pandas/core/generic.py:
```diff
@@ -6081,89 +6081,191 @@ def replace(self, to_replace=None, value=None, inplace=False, limit=None,
+
+        Note how the last entry in column `a` is interpolated differently,
+        because there is no entry after it to use for interpolation.
+        Note how the first entry in column `b` remains `NaN`, because there
```
Double backticks

WillAyd reviewed (changes requested):
IIUC needs some fixes around backtick usage

datapythonista committed:
Fixing quotes and backticks

datapythonista commented:
Yep, agree. I think they should be all right now. Thanks @WillAyd !

WillAyd commented on pandas/core/generic.py:
```diff
@@ -6220,9 +6220,9 @@ def replace(self, to_replace=None, value=None, inplace=False, limit=None,
-        an `order` (int).
+        Filling in ``NaN`` in a Series via polynomial interpolation or splines:
+        Both 'polynomial' and 'spline' methods require that you also specify
+        an ``order`` (int).
```
Arguments should be single backticks - double is only for literals right and code samples I think

datapythonista replied:
My understanding is that double backticks is for code, including parts like a single variable `None`, an assignment `foo=1`... Single backticks is for things that you can refer (link) to, like a function, class, module... And for values just quotes.

For an argument, I'd consider it more code, that something you can link to. That's why I added double backticks. But it's very subtle, I'd be happy with any option (no quoting, single backticks, double backticks and quotes).

Does this make sense?

WillAyd replied:
Yep thanks. I think I've seen other instances where parameters are in single backticks but this is nuanced enough that it shouldn't hold up the PR - can be part of a larger conversation.

datapythonista replied:
I added a bullet point to #20298 to decide a standard for these cases. I think at the moment there is not much consistency.

WillAyd commented:
Thanks @math-and-data and @datapythonista !

khyox commented on pandas/core/generic.py:
```diff
@@ -6081,89 +6081,191 @@ def replace(self, to_replace=None, value=None, inplace=False, limit=None,
+            interpolator).
+        scipy.interpolate.PchipInterpolator : PCHIP 1-d monotonic cubic
+            interpolation.
+        scipy.interpolate.CubicSpline : Cubic spline data interpolator.
```
As this method is referenced here, I expected it to be available just like any of the other ones, but I have combed the source code and I am unable to find the place where this method is used. Was it added because it was planned to add future support for CubicSpline (with a wrapper such as Akima's)?
